In [3]:
import numpy as np

In [4]:
textwidth=30
text2split="Mignonne allons voir si la rose Qui ce matin avoit desclose Sa \
robe de pourpre au Soleil a point perdu ceste vesprée Les plis de sa robe \
pourprée Et son teint au vostre pareil Las voyez comme en peu d'espace \
Mignonne elle a dessus la place Las las ses beautez laissé cheoir O vrayment \
marastre Nature Puis qu'une telle fleur ne dure Que du matin jusques au soir !"
words=text2split.split()

# Recursive_prefixe

In [5]:
def badness(w,taille):
    if len(w)==1:
        return (taille-len(w[0]))**3
    t=sum([len(i) for i in  w])+len(w)-1 
    if t>taille:
        return np.inf
    return (taille-t)**3

In [6]:
def justifiDP_rec(words,i,text_width,seb):
    if i in seb:
        return seb[i]
    mini=np.inf
    for j in range(len(words),i,-1):
        val =badness(words[i:j],text_width)+justifiDP_rec(words,j,text_width,seb)
        if val<=mini:
            mini=val
    seb[i]=mini
    return mini

In [7]:
seb={}
seb[len(words)]=0
justifiDP_rec(words,0,textwidth,seb)
   

897

# Bottom-up

In [8]:
def justifiDP_bottomUp(words,text_width):
    seb={}
    seb[len(words)]=0
    for i in range(len(words)-1,-1,-1):
        mini=np.inf
        for j in range(i+1,len(words)+1):
            val =badness(words[i:j],text_width) + seb[j]
            if val<=mini:
               mini=val
        seb[i]=mini
    return seb[0],seb
    

In [9]:
a,b=justifiDP_bottomUp(words,textwidth)

## Approche greedy

In [10]:
def gredy(words,taille):
    badness=0 
    long_courant=0  #Le nombre de mot courant dans une ligne
    for mot in words:
        if long_courant+len(mot) >taille:
            badness+=(taille-long_courant+1)**3  # +1 pour pour comptabiliser l'espace que j'ai ajouté comme badness puisque on est en fin de ligne
            long_courant=len(mot)
        else:
            long_courant+=len(mot)+1  #+1 pour l'epace entre les mots
    return badness

In [11]:
gredy(words,textwidth)

1333

## Le badness avec approche greedy est plus grand (1333) 

# Exercice 2 – Parenthétisation

In [12]:
d=[13,5,89,3,34]
parent1={}
def init_memoire():
   memo={}
   for i in range(len(d)-1):
        memo[(i,i)]=0
   return memo
memo=init_memoire()

In [13]:
def dp(i,j,parent):
    if (i,j) in memo:
        return  memo[(i,j)]
    mini=np.inf
    for k in range(i,j):
        new=dp(i,k,parent)+dp(k+1,j,parent)+d[i]*d[k+1]*d[j+1]
        if new<mini:
            mini=new
            ind=k
    memo[(i,j)]=mini
    parent[(i,j)]=ind
    return mini



In [14]:
dp(0,3,parent1)

2856

In [15]:
parent1

{(2, 3): 2, (1, 2): 1, (1, 3): 2, (0, 1): 0, (0, 2): 0, (0, 3): 2}

In [16]:
memo

{(0, 0): 0,
 (1, 1): 0,
 (2, 2): 0,
 (3, 3): 0,
 (2, 3): 9078,
 (1, 2): 1335,
 (1, 3): 1845,
 (0, 1): 5785,
 (0, 2): 1530,
 (0, 3): 2856}